File: iOS/Code/Xamarin/bugTrap/bugTrapKit/Controllers/CollectionViewControllers/BtImageCollectionViewController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using CoreGraphics;
using Foundation;
using Photos;
using UIKit;

namespace bugTrapKit
{
	public partial class BtImageCollectionViewController : BtCollectionViewController, IUICollectionViewDelegateFlowLayout, IPHPhotoLibraryChangeObserver
	{
		static nfloat size = (UIScreen.MainScreen.Bounds.Width - 6) / 4;
		static nfloat scale = UIScreen.MainScreen.Scale;

		const string storyboardId = "BtImageCollectionViewController", 
			imageCell = "BtImageCollectionViewCell", addImageCell = "BtAddImageCollectionViewCell";

		CGSize cellSize = CGSize.Empty, thumbSize = CGSize.Empty;
		bool selecting;
		List<NSIndexPath> editIndexes = new List<NSIndexPath> ();
		PHFetchOptions fetchOptions = new PHFetchOptions ();
		PHFetchResult assetsFetchResults;
		Dictionary<NSIndexPath, int> imageRequests = new Dictionary<NSIndexPath, int> ();


		bool selectingSnapshotImagesForExport {
			get {
				var navController = NavigationController as BtImageNavigationController;
				return navController?.SelectingSnapshotImagesForExport ?? false;
			}
			set {
				var navController = NavigationController as BtImageNavigationController;
				if (navController != null) navController.SelectingSnapshotImagesForExport = value;
			}
		}

		public BtImageCollectionViewController (IntPtr handle) : base(handle)
		{
			fetchOptions.SortDescriptors = new [] { new NSSortDescriptor ("modificationDate", false) };

			Task.Run(async() => {
				#if UITEST
				await Task.Delay(3000);
				#endif
				await TrapState.Shared.GetAlbumLocalIdentifier();
				assetsFetchResults = PHAsset.FetchAssets(PHAssetMediaType.Image, fetchOptions);
			});
		}


		public override void ViewDidLoad ()
		{
			base.ViewDidLoad();

			ScreenName = GAIKeys.ScreenNames.SelectSnapshot;

			cellSize = new CGSize (size, size);
			thumbSize = new CGSize (cellSize.Width * scale, cellSize.Height * scale);

			noImageView.Frame = UIScreen.MainScreen.Bounds;

			CollectionView.AllowsMultipleSelection = true;
		}


		public override void ViewWillAppear (bool animated)
		{
			base.ViewWillAppear(animated);

			PHPhotoLibrary.SharedPhotoLibrary.RegisterChangeObserver(this);

			// var navController = NavigationController as BtImageNavigationController;
			// if (navController != null) selecting = navController.SelectingSnapshotImagesForExport;

			selecting = selectingSnapshotImagesForExport;

			updateViewConfiguration(false, updateScreenName: false, reloadData: true);

			var selectedIndexs = CollectionView.GetIndexPathsForSelectedItems();
			if (selectedIndexs != null) {
				foreach (var indexPath in selectedIndexs) {
					CollectionView.DeselectItem(indexPath, false);
				}
			}
		}


		public override void ViewDidDisappear (bool animated)
		{
			PHPhotoLibrary.SharedPhotoLibrary.UnregisterChangeObserver(this);

			base.ViewDidDisappear(animated);
		}


		#region PHPhotoLibraryChangeObserver

		List<NSIndexPath> indexPathsFromIndexSet (NSIndexSet indexSet)
		{
			List<NSIndexPath> indexPaths = new List<NSIndexPath> ();

			if (indexSet == null) return indexPaths;

			indexSet.EnumerateIndexes((nuint idx, ref bool stop) => indexPaths.Add(NSIndexPath.FromItemSection((nint)idx, 0)));

			return indexPaths;
		}


		public void PhotoLibraryDidChange (PHChange changeInstance)
		{
			BeginInvokeOnMainThread(() => {

				// check if there are changes to the assets (insertions, deletions, updates)
				var collectionChanges = changeInstance.GetFetchResultChangeDetails(assetsFetchResults);

				if (collectionChanges != null) {

					assetsFetchResults = collectionChanges.FetchResultAfterChanges;

					if (!collectionChanges.HasIncrementalChanges || collectionChanges.HasMoves) {

						CollectionView.ReloadData();

					} else {
					
						var removedIndexes = indexPathsFromIndexSet(collectionChanges.RemovedIndexes);

						var insertedIndexes = indexPathsFromIndexSet(collectionChanges.InsertedIndexes);

						var changedIndexes = indexPathsFromIndexSet(collectionChanges.ChangedIndexes);

						// if we have incremental diffs, tell the collection view to animate insertions and deletions
						CollectionView.PerformBatchUpdates(() => {

							if (!removedIndexes.IsEmpty()) CollectionView.DeleteItems(removedIndexes.ToArray());

							if (!insertedIndexes.IsEmpty()) CollectionView.InsertItems(insertedIndexes.ToArray());

							if (!changedIndexes.IsEmpty()) CollectionView.ReloadItems(changedIndexes.ToArray());

						}, null);
					}
				}
			});
		}

		#endregion

		public override UIBarButtonItem EditButtonItem {
			get { return editButton; }
		}


		public override nint NumberOfSections (UICollectionView collectionView)
		{
			return 1;
		}


		// nint itemCountForEconfigureation {
		// get { return TrapState.Shared.InActionExtension ? TrapState.Shared.ExtensionSnapshotImageCount : assetsFetchResults?.Count ?? 0; }
		// }

		public override nint GetItemsCount (UICollectionView collectionView, nint section)
		{
			return TrapState.Shared.InActionExtension ? TrapState.Shared.ExtensionSnapshotImageCount : assetsFetchResults?.Count ?? 0;
		}


		public override UICollectionViewCell GetCell (UICollectionView collectionView, NSIndexPath indexPath)
		{
			var cell = collectionView.DequeueReusableCell(imageCell, indexPath) as BtImageCollectionViewCell;

			if (TrapState.Shared.InActionExtension) {

				cell.SetData(TrapState.Shared.GetExtensionSnapshotImageAtIndex(indexPath.Item));

				cell.SetSelected(TrapState.Shared.IsActiveSnapshot(indexPath.Item));

			} else {

				var manager = PHImageManager.DefaultManager;

				if (imageRequests.ContainsKey(indexPath)) manager.CancelImageRequest(imageRequests[indexPath]);

				var asset = assetsFetchResults[indexPath.Item] as PHAsset;

				imageRequests[indexPath] = manager.RequestImageForAsset(asset, thumbSize, PHImageContentMode.AspectFill, null, 
					(result, info) => cell.SetData(result)
				);
			}

			return cell;
		}


		public override void ItemSelected (UICollectionView collectionView, NSIndexPath indexPath)
		{
			if (TrapState.Shared.InActionExtension) {
				SaveActiveSnapshotAndResetAnnotateImageView(indexPath.Item, null);
				return;
			}

			if (Editing && !editIndexes.Contains(indexPath)) {

				editIndexes.Add(indexPath);

			} else if (selecting) { // selecting images to add to an Issue

				var asset = assetsFetchResults?[indexPath.Item] as PHAsset;

				TrapState.Shared.AddSnapshotImage(asset.LocalIdentifier);

			} else {

				var asset = assetsFetchResults?[indexPath.Item] as PHAsset;

				var selectedItems = CollectionView.GetIndexPathsForSelectedItems().Where(ip => ip != indexPath);
				CollectionView.PerformBatchUpdates(() => {
					foreach (var item in selectedItems) {
						CollectionView.DeselectItem(item, true);
					}
				}, null);

				SaveActiveSnapshotAndResetAnnotateImageView(null, asset?.LocalIdentifier);
			}
		}

		public override bool ShouldSelectItem (UICollectionView collectionView, NSIndexPath indexPath)
		{
			return true;
		}

		public override bool ShouldDeselectItem (UICollectionView collectionView, NSIndexPath indexPath)
		{
			var deselect = Editing || selecting;

			if (deselect) SaveActiveSnapshotAndResetAnnotateImageView(null, TrapState.Shared.LocalIdentifiersForActiveSnapshot);

			return deselect;
		}


		public override void ItemDeselected (UICollectionView collectionView, NSIndexPath indexPath)
		{
			if (Editing) {

				editIndexes.Remove(indexPath);

			} else if (selecting) {

				var asset = assetsFetchResults?[indexPath.Item] as PHAsset;

				TrapState.Shared.RemoveSnapshotImage(asset?.LocalIdentifier);

				// may need to accout for this view presented over the new bug details tvc
				// TrackerService.Shared.setCurrentTrackerType(.None) // why?

			} else {

				var asset = assetsFetchResults?[indexPath.Item] as PHAsset;

				SaveActiveSnapshotAndResetAnnotateImageView(null, asset?.LocalIdentifier);
			}
		}


		[Export("collectionView:layout:sizeForItemAtIndexPath:")]
		public CGSize GetSizeForItem (UICollectionView collectionView, UICollectionViewLayout layout, NSIndexPath indexPath)
		{
			return cellSize;
		}


		void SaveActiveSnapshotAndResetAnnotateImageView (nint? indexForExtension, string localIdentifier)
		{
			// var navController = NavigationController as BtImageNavigationController;
			// if (navController != null) navController.SelectingSnapshotImagesForExport = false;

			selectingSnapshotImagesForExport = false;

			var annotateImageNavController = PresentingViewController?.ChildViewControllers?[0] as BtAnnotateImageNavigationController;

			var annotateImageController = annotateImageNavController?.TopViewController as BtAnnotateImageViewController ??
			                              PresentingViewController?.ChildViewControllers?[0] as BtAnnotateImageViewController;

			annotateImageController.SaveCurrentSnapshotAndResetState(indexForExtension, localIdentifier);

			DismissViewController(true, null);
		}



		partial void titleClicked (UIButton sender)
		{
			editIndexes = new List<NSIndexPath> ();

			Editing = false;

			updateViewConfiguration(false, selecting);

			TrapState.Shared.ResetSnapshotImages(false);

			selectingSnapshotImagesForExport = false;

			DismissViewController(true, null);
		}


		partial void cancelClicked (UIBarButtonItem sender)
		{
			// cancel should only appear if selecting screenshots for a new bug, or in edit state (delete)

			if (Editing) {

				Editing = false;

				updateViewConfiguration();

				CollectionView.ReloadItems(editIndexes.ToArray());

				editIndexes = new List<NSIndexPath> ();
			
			} else if (selecting) {

				TrapState.Shared.ResetSnapshotImages();

				selectingSnapshotImagesForExport = false;

				DismissViewController(true, null);
			}
		}


		async partial void trashClicked (UIBarButtonItem sender)
		{
			Editing = false;

			updateViewConfiguration(showActivity: true);

			var editAssets = editIndexes.Select(ip => assetsFetchResults[ip.Item] as PHAsset);

			await TrapState.Shared.DeleteAssets(editAssets.ToList());

			editIndexes = new List<NSIndexPath> ();

			BeginInvokeOnMainThread(() => updateViewConfiguration());
		}


		partial void editClicked (UIBarButtonItem sender)
		{
			Editing = true;

			updateViewConfiguration();
		}


		async partial void doneClicked (UIBarButtonItem sender)
		{
			if (selecting) {

				selectingSnapshotImagesForExport = false;

				var annotateImageNavController = PresentingViewController.ChildViewControllers?.FirstOrDefault() as BtAnnotateImageNavigationController ?? PresentingViewController as BtAnnotateImageNavigationController;

				await DismissViewControllerAsync(true);

				annotateImageNavController?.PresentNewBugDetailsNavController(true);
			
			} else {
				
				DismissViewController(true, null);
			}
		}


		partial void addClicked (UIBarButtonItem sender)
		{
			
		}


		void updateViewConfiguration (bool animate = true, bool reloadData = false, bool updateScreenName = true, bool showActivity = false)
		{
			if (GetItemsCount(CollectionView, 0) == 0 && !noImageView.IsDescendantOfView(View)) {
				View.AddSubview(noImageView);
			} else if (noImageView.IsDescendantOfView(View)) {
				noImageView.RemoveFromSuperview();
			}

			if (updateScreenName) UpdateScreenName(Editing ? GAIKeys.ScreenNames.EditSnapshots : GAIKeys.ScreenNames.SelectSnapshot);

			CollectionView.AllowsMultipleSelection = true;

			var leftItem = selecting || Editing ? cancelButton : editButton;
			var rightItem = showActivity ? activityButton : Editing ? trashButton : selecting ? doneButton : null;

			if (NavigationItem.LeftBarButtonItem != leftItem) NavigationItem.SetLeftBarButtonItem(leftItem, animate);
			if (NavigationItem.RightBarButtonItem != rightItem) NavigationItem.SetRightBarButtonItem(rightItem, animate);

			if (reloadData) {

				if (TrapState.Shared.InActionExtension) {

					CollectionView.ReloadData();

				} else {
					
					Task.Run(async () => {
						
						var oldCount = assetsFetchResults?.Count ?? 0;

						await TrapState.Shared.GetAlbumLocalIdentifier();

						PHPhotoLibrary.SharedPhotoLibrary.UnregisterChangeObserver(this);

						assetsFetchResults = PHAsset.FetchAssets(PHAssetMediaType.Image, fetchOptions);

						if ((assetsFetchResults?.Count ?? 0) > oldCount) InvokeOnMainThread(() => CollectionView.ReloadData());

						PHPhotoLibrary.SharedPhotoLibrary.RegisterChangeObserver(this);
					});
				}
			}
		}
		//		} else if let oldCount = assetsFetchResults?.count {
		//
		//			TrapState.Shared.getAlbumLocalIdentifier() { _ in
		//
		//				PHPhotoLibrary.sharedPhotoLibrary().unregisterChangeObserver(self)
		//
		//				self.assetsFetchResults = PHAsset.fetchAssetsWithMediaType(.Image, options: self.fetchOptions)
		//
		//				if self.assetsFetchResults?.count > oldCount {
		//
		//					dispatch_async(dispatch_get_main_queue()) {
		//
		//						self.collectionView!.reloadData()
		//					}
		//				}
		//
		//				PHPhotoLibrary.sharedPhotoLibrary().registerChangeObserver(self)
		//			}
		//		}
		// }



		public override bool PrefersStatusBarHidden () => true;

		public override UIStatusBarStyle PreferredStatusBarStyle () => UIStatusBarStyle.LightContent;
	}
}

File: iOS/Code/Swift/bugTrap/bugTrapKit/Controllers/CollectionViewControllers/BtImageCollectionViewController.swift
//
//  BtImageCollectionViewController.swift
//  bugTrap
//
//  Created by Colby L Williams on 10/28/14.
//  Copyright (c) 2014 bugTrap. All rights reserved.
//

import UIKit
import Photos

let reuseIdentifier = "Cell"

class BtImageCollectionViewController: BtCollectionViewController, UICollectionViewDelegateFlowLayout, PHPhotoLibraryChangeObserver {

	let size = (UIScreen.mainScreen().bounds.width - 6) / 4
	
	let scale = UIScreen.mainScreen().scale
	
	let storyboardId = "BtImageCollectionViewController", imageCell = "BtImageCollectionViewCell", addImageCell = "BtAddImageCollectionViewCell"
	
	@IBOutlet var activityButton: UIBarButtonItem!
	
	@IBOutlet var cancelButton: UIBarButtonItem!
	
	@IBOutlet var trashButton: UIBarButtonItem!
	
	@IBOutlet var titleButton: UIButton!
	
	@IBOutlet var editButton: UIBarButtonItem!
	
	@IBOutlet var doneButton: UIBarButtonItem!
	
	@IBOutlet var addButton: UIBarButtonItem!
	
	@IBOutlet var noImageDetail: UILabel!
	@IBOutlet var noImageTitle: UILabel!
	@IBOutlet var noImageView: UIView!
	
	
	var cellSize = CGSize.zeroSize, thumbSize = CGSize.zeroSize
	
	var selecting = false
	
	var editIndexes: [NSIndexPath] = []
	
	var fetchOptions = PHFetchOptions()
	
	var assetsFetchResults: PHFetchResult?
	
	var imageRequests: [NSIndexPath: PHImageRequestID] = [:]
	
	required init?(coder aDecoder: NSCoder) {
		super.init(coder: aDecoder)
		
		fetchOptions.sortDescriptors = [ NSSortDescriptor(key: "modificationDate", ascending: false) ]
		
		TrapState.Shared.getAlbumLocalIdentifier() { _ in
			self.assetsFetchResults = PHAsset.fetchAssetsWithMediaType(.Image, options: self.fetchOptions)
			// PHPhotoLibrary.sharedPhotoLibrary().registerChangeObserver(self)
		}
	}
	
	
	override func viewDidLoad() {
		super.viewDidLoad()
		
		screenName = GAIKeys.ScreenNames.selectSnapshot
		
		cellSize = CGSize(width: size, height: size)
		
		thumbSize = CGSize(width: cellSize.width * scale, height: cellSize.height * scale)
		
		noImageView.frame = UIScreen.mainScreen().bounds
		
		collectionView?.allowsMultipleSelection = true
	}
	
	
	
	override func viewWillAppear(animated: Bool) {
		super.viewWillAppear(animated)
		
		PHPhotoLibrary.sharedPhotoLibrary().registerChangeObserver(self)
		
		if let navController = navigationController as? BtImageNavigationController {
			selecting = navController.selectingSnapshotImagesForExport
		}
		
		updateViewConfiguration(false, updateScreenName: false, reloadData: true)

		if let selectedIndexes = collectionView?.indexPathsForSelectedItems() {
			for index in selectedIndexes {
				collectionView?.deselectItemAtIndexPath(index as NSIndexPath, animated: false)
			}
		}
	}
	
	
	override func viewDidDisappear(animated: Bool) {
		super.viewDidDisappear(animated)

		PHPhotoLibrary.sharedPhotoLibrary().unregisterChangeObserver(self)
	}
	
	
	// MARK: PHPhotoLibraryChangeObserver
	
	func indexPathsFromIndexSet(indexSet: NSIndexSet) -> [NSIndexPath] {
		
		var indexPaths = [NSIndexPath]()

		indexSet.enumerateIndexesUsingBlock { idx, _ in
			indexPaths.append(NSIndexPath(forItem: idx, inSection: 0))
		}
		
		return indexPaths
	}

	
	func photoLibraryDidChange(changeInstance: PHChange) {
		
		dispatch_async(dispatch_get_main_queue()) {

			if let assets = self.assetsFetchResults {
			
				// check if there are changes to the assets (insertions, deletions, updates)
				if let collectionChanges = changeInstance.changeDetailsForFetchResult(assets) {

					self.assetsFetchResults = collectionChanges.fetchResultAfterChanges
					
					// get the new fetch result
					if !collectionChanges.hasIncrementalChanges || collectionChanges.hasMoves {

						self.collectionView?.reloadData()
						
					} else { //if collectionChanges.hasIncrementalChanges {

						var removedIndexes	= [NSIndexPath]()
						var insertedIndexes = [NSIndexPath]()
						var changedIndexes	= [NSIndexPath]()
						
						if let removed = collectionChanges.removedIndexes {
							removedIndexes = self.indexPathsFromIndexSet(removed)
						}
						
						if let inserted = collectionChanges.insertedIndexes {
							insertedIndexes = self.indexPathsFromIndexSet(inserted)
						}
						
						if let changed = collectionChanges.changedIndexes {
							changedIndexes = self.indexPathsFromIndexSet(changed)
						}
						
						// if we have incremental diffs, tell the collection view to animate insertions and deletions
						self.collectionView?.performBatchUpdates({
							
							if !removedIndexes.isEmpty {
								self.collectionView?.deleteItemsAtIndexPaths(removedIndexes)
							}
							
							if !insertedIndexes.isEmpty {
								self.collectionView?.insertItemsAtIndexPaths(insertedIndexes)
							}
							
							if !changedIndexes.isEmpty {
								self.collectionView?.reloadItemsAtIndexPaths(changedIndexes)
							}
							
						}) { _ in
						
						}
					}
				}
			}
		}
	}
	
	
	
	override func editButtonItem() -> UIBarButtonItem {
		return editButton
	}
	
	
	
	override func numberOfSectionsInCollectionView(collectionView: UICollectionView) -> Int {
		return 1
	}

	
	var itemCountForConfiguration: Int {
		
		var resultsCount = 0, extensionSnapshotImageCount = TrapState.Shared.getExtensionSnapshotImageCount()
		
		if let results = assetsFetchResults {
			resultsCount = results.count
		}
		
		return TrapState.Shared.inActionExtension ? extensionSnapshotImageCount : resultsCount
	}
	
	
	override func collectionView(collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
		
		return itemCountForConfiguration
	}
	
	
	
	override func collectionView(collectionView: UICollectionView, cellForItemAtIndexPath indexPath: NSIndexPath) -> UICollectionViewCell {
		
		let cell = collectionView.dequeueReusableCellWithReuseIdentifier(imageCell, forIndexPath: indexPath) as! BtImageCollectionViewCell
		
		if TrapState.Shared.inActionExtension {
		
			cell.setData(TrapState.Shared.getExtensionSnapshotImageAtIndex(indexPath.item))
			
			cell.setIsSelected(TrapState.Shared.isActiveSnapshot(indexPath.item))

		} else {
		
			let manager = PHImageManager.defaultManager()
		
			if let request = imageRequests[indexPath] {
				manager.cancelImageRequest(request)
			}
		
			let asset = assetsFetchResults![indexPath.item] as! PHAsset
		
			imageRequests[indexPath] = manager.requestImageForAsset(asset, targetSize: thumbSize, contentMode: .AspectFill, options: nil) { (result, _) in
				cell.setData(result)
			}

			// cell.selected = TrapState.Shared.isSelectedSnapshot(asset.localIdentifier)
			
			// cell.setSelected(TrapState.Shared.isSelectedSnapshot(asset.localIdentifier))
		}
		
		return cell
	}
	

	
	override func collectionView(collectionView: UICollectionView, didSelectItemAtIndexPath indexPath: NSIndexPath) {
		
		// Log.debug(self, "didSelect")
		
		if TrapState.Shared.inActionExtension {
			self.saveActiveSnapshotAndResetAnnotateImageView(indexPath.item, localIdentifier: nil)
			return
		}
		
		if editing && !editIndexes.contains(indexPath) {
			
			editIndexes.append(indexPath)

		} else if selecting { // selecting images to add to an Issue
			
			let asset = assetsFetchResults![indexPath.item] as! PHAsset
			
			TrapState.Shared.addSnapshotImage(asset.localIdentifier)
			
		} else {
			
			let asset = assetsFetchResults![indexPath.item] as! PHAsset
			
			if let selectedItems = collectionView.indexPathsForSelectedItems()?.filter({ $0 as NSIndexPath != indexPath }) {
				collectionView.performBatchUpdates({
					for item in selectedItems {
						collectionView.deselectItemAtIndexPath(item, animated: true)
					}
				}, completion: nil)
			}
			
			saveActiveSnapshotAndResetAnnotateImageView(nil, localIdentifier: asset.localIdentifier)
		}
	}
	
	override func collectionView(collectionView: UICollectionView, shouldSelectItemAtIndexPath indexPath: NSIndexPath) -> Bool {
		// Log.debug(self, "shouldSelect")
		return true
	}
	
	override func collectionView(collectionView: UICollectionView, shouldDeselectItemAtIndexPath indexPath: NSIndexPath) -> Bool {

		// Log.debug(self, "shouldDeselect")
		
		let deselect = editing || selecting
		
		if !deselect {
			saveActiveSnapshotAndResetAnnotateImageView(nil, localIdentifier: TrapState.Shared.getLocalIdentifierForActiveSnapshot())
		}
		
		return deselect
	}
	
	
	override func collectionView(collectionView: UICollectionView, didDeselectItemAtIndexPath indexPath: NSIndexPath) {
		
		// Log.debug(self, "didDeselect")
		
		if editing {
			
			if let index = editIndexes.indexOf(indexPath) {
				editIndexes.removeAtIndex(index)
			}
			
		} else if selecting { // selecting images to add to an Issue
			
			let asset = assetsFetchResults![indexPath.item] as! PHAsset
			
			TrapState.Shared.removeSnapshotImage(asset.localIdentifier)
		
			// may need to accout for this view presented over the new bug details tvc
			// TrackerService.Shared.setCurrentTrackerType(.None) // why?
			
		} else {
			
			let asset = assetsFetchResults![indexPath.item] as! PHAsset
			
			saveActiveSnapshotAndResetAnnotateImageView(nil, localIdentifier: asset.localIdentifier)
		}
	}
	

	
	func collectionView(collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAtIndexPath indexPath: NSIndexPath) -> CGSize {
		return cellSize
	}
	
	
	
	func saveActiveSnapshotAndResetAnnotateImageView (indexForExtension: Int?, localIdentifier: String?) {
		
		if let navController = navigationController as? BtImageNavigationController {
			navController.selectingSnapshotImagesForExport = false
		}
		
		if let annotateImageNavController = presentingViewController?.childViewControllers[0] as? BtAnnotateImageNavigationController {
			
			if let annotateImageController = annotateImageNavController.topViewController as? BtAnnotateImageViewController {
				annotateImageController.saveCurrentSnapshotAndResetState(indexForExtension, localIdentifierOfActiveSnapshot: localIdentifier)
			}
			
		} else if let annotateImageController = presentingViewController?.childViewControllers[0] as? BtAnnotateImageViewController {
			annotateImageController.saveCurrentSnapshotAndResetState(indexForExtension, localIdentifierOfActiveSnapshot: localIdentifier)
		}
		
		dismissViewControllerAnimated(true, completion: nil)
	}
	
	
	
	@IBAction func titleClicked(sender: UIButton) {

		editIndexes = []
		
		editing = false;
		
		updateViewConfiguration(false, reloadData: selecting)
		
		TrapState.Shared.resetSnapshotImages(false)
		
		if let navController = navigationController as? BtImageNavigationController {
			navController.selectingSnapshotImagesForExport = false
		}
		
		dismissViewControllerAnimated(true, completion: nil)
	}
	
	
	@IBAction func cancelClicked(sender: UIBarButtonItem) {
		
		// cancel should only appear if selecting screenshots for a new bug, or in edit state (delete)
		
		if editing {
			
			editing = false
			
			updateViewConfiguration()
			
			collectionView?.reloadItemsAtIndexPaths(editIndexes)
			
			editIndexes = []
			
		} else if selecting {
			
			TrapState.Shared.resetSnapshotImages()
			
			if let navController = navigationController as? BtImageNavigationController {
				navController.selectingSnapshotImagesForExport = false
			}
			
			dismissViewControllerAnimated(true, completion: nil)
		}
	}
	
	
	@IBAction func trashClicked(sender: UIBarButtonItem) {
		
		editing = false
		
		updateViewConfiguration(showActivity: true)
		
		let editAssets = editIndexes.map({ self.assetsFetchResults![$0.item] as! PHAsset })
		
		TrapState.Shared.deleteAssets(editAssets) {
			self.editIndexes = []
			dispatch_async(dispatch_get_main_queue()) {
				self.updateViewConfiguration()
			}
		}
	}

	
	@IBAction func editClicked(sender: UIBarButtonItem) {
		
		editing = true
		
		updateViewConfiguration()
	}
	
	
	@IBAction func doneClicked(sender: UIBarButtonItem) {
		
		if selecting {
			
			if let navController = navigationController as? BtImageNavigationController {
				navController.selectingSnapshotImagesForExport = false
			}

			if let annotateImageNavController = presentingViewController?.childViewControllers[0] as? BtAnnotateImageNavigationController {
				dismissViewControllerAnimated(true) {
					annotateImageNavController.presentNewBugDetailsNavController(true, block: nil)
				}
			} else if let annotateImageNavController = presentingViewController as? BtAnnotateImageNavigationController {
				dismissViewControllerAnimated(true) {
					annotateImageNavController.presentNewBugDetailsNavController(true, block: nil)
				}
			}
			
		} else {
			
			dismissViewControllerAnimated(true, completion: nil)
		}
	}
	
	
	@IBAction func addClicked(sender: UIBarButtonItem) {
		
//		adding = true

//		updateViewConfiguration(reloadData: true)
		
		//presentViewController(libraryNavigationController, animated: true, completion: nil)
	}
	
	
	func updateViewConfiguration(animate: Bool = true, reloadData: Bool = false, updateScreenName: Bool = true, showActivity: Bool = false) {
		
//		if itemCountForConfiguration == 0 && !noImageView.isDescendantOfView(view){
		if itemCountForConfiguration == 0 && !noImageView.isDescendantOfView(view){
			view.addSubview(noImageView)
		} else if noImageView.isDescendantOfView(view) {
			noImageView.removeFromSuperview()
		}
		
		if updateScreenName {
			self.updateScreenName(editing ? GAIKeys.ScreenNames.editSnapshots : GAIKeys.ScreenNames.selectSnapshot)
		}
		
		collectionView?.allowsMultipleSelection = true
		
		let leftItem =	selecting || editing ? cancelButton : editButton
		let rightItem = showActivity ? activityButton : editing ? trashButton : selecting ? doneButton : nil // addButton
		
		if navigationItem.leftBarButtonItem != leftItem {
			navigationItem.setLeftBarButtonItem(leftItem, animated: animate)
		}
		
		if navigationItem.rightBarButtonItem != rightItem {
			navigationItem.setRightBarButtonItem(rightItem, animated: animate)
		}
		
		if reloadData {
		
			if TrapState.Shared.inActionExtension {
				
				collectionView?.reloadData()
			
			} else if let oldCount = assetsFetchResults?.count {
				
				TrapState.Shared.getAlbumLocalIdentifier() { _ in
					
					PHPhotoLibrary.sharedPhotoLibrary().unregisterChangeObserver(self)
					
					self.assetsFetchResults = PHAsset.fetchAssetsWithMediaType(.Image, options: self.fetchOptions)
					
					if self.assetsFetchResults?.count > oldCount {
						
						dispatch_async(dispatch_get_main_queue()) {
							
							self.collectionView!.reloadData()
						}
					}
					
					PHPhotoLibrary.sharedPhotoLibrary().registerChangeObserver(self)
				}
			}
		}

//		} else if let oldCount = assetsFetchResults?.count {
//			
//			TrapState.Shared.getAlbumLocalIdentifier() { _ in
//				
//				PHPhotoLibrary.sharedPhotoLibrary().unregisterChangeObserver(self)
//				
//				self.assetsFetchResults = PHAsset.fetchAssetsWithMediaType(.Image, options: self.fetchOptions)
//				
//				if self.assetsFetchResults?.count > oldCount {
//					
//					dispatch_async(dispatch_get_main_queue()) {
//						
//						self.collectionView!.reloadData()
//					}
//				}
//				
//				PHPhotoLibrary.sharedPhotoLibrary().registerChangeObserver(self)
//			}
//		}
	}
	
	
	override func prefersStatusBarHidden() -> Bool {
		return true
	}
	
	override func preferredStatusBarStyle() -> UIStatusBarStyle {
		return .LightContent
	}
	
    // MARK: UICollectionViewDelegate
}
